source/common/http/http2/codec_impl.cc

```diff
@@ -374,17 +419,57 @@ void ConnectionImpl::StreamImpl::readDisable(bool disable) {
 }
 
 void ConnectionImpl::StreamImpl::pendingRecvBufferHighWatermark() {
-  ENVOY_CONN_LOG(debug, "recv buffer over limit ", parent_.connection_);
-  ASSERT(!pending_receive_buffer_high_watermark_called_);
-  pending_receive_buffer_high_watermark_called_ = true;
-  readDisable(true);
+  if (!Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.defer_processing_backedup_streams")) {
+    ENVOY_CONN_LOG(debug, "recv buffer over limit ", parent_.connection_);
+    ASSERT(!pending_receive_buffer_high_watermark_called_);
+    pending_receive_buffer_high_watermark_called_ = true;
+    readDisable(true);
+  }
 }
 
 void ConnectionImpl::StreamImpl::pendingRecvBufferLowWatermark() {
-  ENVOY_CONN_LOG(debug, "recv buffer under limit ", parent_.connection_);
-  ASSERT(pending_receive_buffer_high_watermark_called_);
-  pending_receive_buffer_high_watermark_called_ = false;
-  readDisable(false);
+  if (!Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.defer_processing_backedup_streams")) {
+    ENVOY_CONN_LOG(debug, "recv buffer under limit ", parent_.connection_);
+    ASSERT(pending_receive_buffer_high_watermark_called_);
+    pending_receive_buffer_high_watermark_called_ = false;
+    readDisable(false);
+  }
+}
+
+absl::optional<ConnectionImpl::StreamImpl::BackedUpStreamManager::Stage>
+ConnectionImpl::StreamImpl::BackedUpStreamManager::getNextStage() {
+  if (body_buffered_) {
+    return Stage::Body;
+  }
+  if (trailers_buffered_) {
+    return Stage::Trailers;
+  }
+
+  return absl::nullopt;
+}
+
+void ConnectionImpl::StreamImpl::decodeData() {
+  if (Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.defer_processing_backedup_streams") &&
+      buffersOverrun() && !stream_manager_.flush_all_data_) {
+    ENVOY_CONN_LOG(trace, "Stream {} buffering decodeData() call.", parent_.connection_,
+                   stream_id_);
+    stream_manager_.body_buffered_ = true;
+    return;
+  }
+
+  // Any buffered data will be consumed.
+  stream_manager_.body_buffered_ = false;
+
+  // It's possible that we are waiting to send a deferred reset, so only raise data if local
+  // is not complete.
+  if (!deferred_reset_) {
+    decoder().decodeData(*pending_recv_data_, remote_end_stream_);
+  }
+
+  pending_recv_data_->drain(pending_recv_data_->length());
```

Comment:
I thought we were going to dump in chunks, whatever the default read limit is?

Reply:
Ok, let's definitely add a TODO here then.  I think delaying this for a bit is fine but I'd definitely like to sort out the stream flush-close issue before we get too deep into the weeds

Reply:
added todo